src/expr/src/linear.rs

```diff
@@ -1881,6 +1882,132 @@ pub mod plan {
         }
     }
+
+    /// Logic for turning an [MfpPlan] into filters that can confidently skip
+    /// fetching an entire persist part based on aggregate statistics (e.g.
+    /// min/max) about the data contained in it.
+    #[derive(Debug)]
+    pub struct MfpPushdown {
+        relation_desc: RelationDesc,
+        // TODO(mfp): Probably some preprocessing we could do to make
+        // `should_fetch` calls more efficient.
+        mfp: Option<MfpPlan>,
+    }
+
+    impl MfpPushdown {
+        /// Returns a new [MfpPushdown] for the given [MfpPlan] and support
+        /// info.
+        pub fn new(
+            relation_desc: RelationDesc,
+            mfp: Option<&MfpPlan>,
+            // TODO: timestamp/as_of and until.
+        ) -> Self {
+            MfpPushdown {
+                relation_desc,
+                mfp: mfp.cloned(),
+            }
+        }
+
+        /// Returns `false` if a chunk of data with the given statistics would
+        /// entirely be filtered out by this MFP once fetched.
+        ///
+        /// Returns `true` if we don't know.
+        pub fn should_fetch<S: PersistSourceDataStats>(&mut self, stats: &S) -> bool {
+            let mfp = match self.mfp.as_ref() {
+                Some(x) => x,
+                None => return true,
+            };
+            for (support, predicate) in mfp.mfp.predicates.iter() {
+                if mfp.mfp.input_arity < *support {
+                    // We'd need the result of map to evaluate this predicate.
+                    return true;
+                }
+                match self.provably_not_in_part(stats, predicate) {
+                    // The predicates are ANDed, so if any of them are provably
+                    // not in the part, we can confidently skip fetching it.
+                    Some(true) => return false,
+                    // TODO: There's probably some useful distinctions we could
+                    // pass back to persist for tracking: e.g. part definitely
+                    // is not needed vs part definitely might be needed vs we
+                    // don't know because some stats were missing vs we don't
+                    // know because the code doesn't yet handle some expression.
+                    Some(false) | None => continue,
+                }
+            }
+            // TODO: Do the same for self.map.{lower,upper}_bounds.
+            true
+        }
+
+        fn provably_not_in_part<S: PersistSourceDataStats>(
+            &self,
+            stats: &S,
+            expr: &MirScalarExpr,
+        ) -> Option<bool> {
+            let arena = RowArena::default();
+            // TODO: This is entirely just a placeholder impl so that we can get
+            // at least one query to work end-to-end. The real structure would
+            // certainly look different than this.
+            //
+            // It works by sniffing out expressions that (at a high level) look
```

Comment:
Hmm, yes! As a long-term goal, maybe after fixing https://github.com/MaterializeInc/materialize/issues/18091, we could merge this with `ColumnKnowledge` somehow.

Reply:
The point is that in this case you need `mz_expr` to depend on `mz_transform` (so you can call in this method instead of the bespoke code that written below as a first approximation), but in reality the dependency is from `mz_expr` to `mz_transform`.

We'll need to figure out a way to factor out the common bits when we work on that (or move this code out of `linear.rs`).